File: solutions/cpp/protein-translation/1/protein_translation.cpp
#include "protein_translation.h"
#include <string>
#include <vector>
#include <unordered_map>

namespace protein_translation {

std::vector<std::string> proteins(const std::string& rna) {
    std::vector<std::string> result;
    
    // Create codon to protein mapping
    std::unordered_map<std::string, std::string> codon_map = {
        {"AUG", "Methionine"},
        {"UUU", "Phenylalanine"}, {"UUC", "Phenylalanine"},
        {"UUA", "Leucine"}, {"UUG", "Leucine"},
        {"UCU", "Serine"}, {"UCC", "Serine"}, 
        {"UCA", "Serine"}, {"UCG", "Serine"},
        {"UAU", "Tyrosine"}, {"UAC", "Tyrosine"},
        {"UGU", "Cysteine"}, {"UGC", "Cysteine"},
        {"UGG", "Tryptophan"},
        {"UAA", "STOP"}, {"UAG", "STOP"}, {"UGA", "STOP"}
    };
    
    // Process RNA in chunks of 3 characters (codons)
    for (size_t i = 0; i < rna.length(); i += 3) {
        if (i + 3 > rna.length()) {
            break; // Incomplete codon at the end
        }
        
        std::string codon = rna.substr(i, 3);
        
        // Check if codon is in our map
        auto it = codon_map.find(codon);
        if (it == codon_map.end()) {
            // Unknown codon - for this exercise, we'll assume it's not in the input
            // but we could throw an exception if needed
            break;
        }
        
        std::string protein = it->second;
        
        // Check for STOP codon
        if (protein == "STOP") {
            break;
        }
        
        // Add the protein to our result
        result.push_back(protein);
    }
    
    return result;
}

}  // namespace protein_translation

File: solutions/python/protein-translation/1/protein_translation.py
"""Module providing a function to translate RNA sequences into proteins."""


def proteins(strand):
    """
    Translate an RNA strand into a sequence of protein names.
    
    The function reads the RNA strand in groups of three nucleotides (codons)
    and translates each codon to its corresponding amino acid. Translation
    stops when a STOP codon is encountered.
    
    Args:
        strand (str): A string representing an RNA sequence.
        
    Returns:
        list: A list of protein names in the order they are translated.
        
    Examples:
        >>> proteins("AUGUUUUCU")
        ['Methionine', 'Phenylalanine', 'Serine']
        
        >>> proteins("AUGUUUUCUUAAAUG")
        ['Methionine', 'Phenylalanine', 'Serine']
    """
    # Define the codon to amino acid mapping
    codon_to_protein = {
        "AUG": "Methionine",
        "UUU": "Phenylalanine",
        "UUC": "Phenylalanine",
        "UUA": "Leucine",
        "UUG": "Leucine",
        "UCU": "Serine",
        "UCC": "Serine",
        "UCA": "Serine",
        "UCG": "Serine",
        "UAU": "Tyrosine",
        "UAC": "Tyrosine",
        "UGU": "Cysteine",
        "UGC": "Cysteine",
        "UGG": "Tryptophan",
        "UAA": "STOP",
        "UAG": "STOP",
        "UGA": "STOP",
    }
    
    translated_proteins = []
    
    # Process the strand in chunks of 3 nucleotides
    for index in range(0, len(strand), 3):
        codon = strand[index:index + 3]
        
        # Only process complete codons
        if len(codon) < 3:
            break
            
        # Get the amino acid for this codon
        amino_acid = codon_to_protein.get(codon)
        
        # If codon is invalid, we could raise an error or skip
        # For this exercise, we'll skip invalid codons
        if amino_acid is None:
            continue
            
        # Stop translation at STOP codons
        if amino_acid == "STOP":
            break
            
        translated_proteins.append(amino_acid)
    
    return translated_proteins